### PR TITLE
feat(durability): Add documentation for the feature

### PR DESCRIPTION
**Pull request by ieQu1**

### CODE REVIEW

File: en_US/durability/introduction.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+Storage requirements can be estimated according to the following formulae:
+
+- Storage utilized by the messages is proportional to the rate of incoming messages multiplied by the value of configuration parameter `session_persistence.message_retention_period`.
+
+- Storage utilized by the persistent sessions is proportional to the number of sessions multiplied by the number of streams that the sessions subscribe to.
+
+- Number of streams is proportional to the number of shards.
+
+## Message retention
```
Comment (wivwiv) [inline en_US/durability/introduction.md]: Use the table to list all the configuration items, as well as the description of the configuration items

File: dir.yaml [hunk 2]
```diff
@@ -561,6 +561,9 @@
         - file-transfer/quick-start
         - file-transfer/broker
         - file-transfer/client
+    - title_en: Durable MQTT Sessions and Messages
+      title_cn: 持久的MQTT会话和信息
```
Comment (wivwiv) [inline dir.yaml]:
```suggestion
      title_cn: MQTT 会话与消息持久化
```

**Review by wivwiv — CHANGES_REQUESTED** [review]

File: en_US/durability/introduction.md (hunk 1 above)
Comment (Meggielqk) [reply en_US/durability/introduction.md]: Maybe adding a configuration code example with some explanations and what the effect will be after the configuration? Like with an MQTTX pub/sub example? Not sure whether it is necessary at this alpha stage. @wivwiv @ieQu1

Comment (Meggielqk) [reply en_US/durability/introduction.md]: Another question is that don't we need to mention the built-in backend "RocksDB" at all in the document?

Comment (ieQu1) [reply en_US/durability/introduction.md]: > Another question is that don't we need to mention the built-in backend "RocksDB" at all in the document?

Good point. I think it's worth mentioning.

File: en_US/durability/durability_introduction.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+#### Stream
+
+Messages in each shard and generation are split into streams. Streams serve as units of message serialization in EMQX. Streams can contain messages from multiple topics. Various storage layouts can employ different strategies for mapping topics into streams.
+
+Persistent sessions interact with this structure by fetching messages in batches from the streams, with the batch size adjustable via the `session_persistence.max_batch_size` parameter. This comprehensive system ensures that EMQX can efficiently handle durable message storage and retrieval.
+
+### Session Persistence Across Cluster
+
+Each node within an EMQX cluster is assigned a unique site ID, which serves as a stable identifier, independent of the Erlang node name (`emqx@...`). Site IDs are persistent, and they are randomly generated at the first startup of the node. This stability is crucial for maintaining the integrity of persistent sessions and messages, especially in scenarios where nodes might undergo name modifications or reconfigurations.
```
Comment (ieQu1) [inline en_US/durability/durability_introduction.md]: > This stability is crucial for maintaining the integrity of persistent sessions and messages, especially in scenarios where nodes might undergo name modifications or reconfigurations.

This sentence sounds as if the user must somehow worry about stability of the site IDs. In fact, as long as users don't delete data directory, site ID should remain static (as opposed to regular EMQX node names that change if hostname/IP changes), so the intention of this passage is the opposite: don't worry about changing node host names or IPs: the data should remain readable. 

So maybe rephrase it in a more informational style (remove "crucial")?

File: en_US/durability/durability_introduction.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
+# Durable MQTT Sessions and Messages
+
+Starting from release v5.5.x, EMQX contains an embedded durable storage for MQTT sessions and messages. This page introduces the session persistence feature EMQX and how it's capability ensures the resumption of sessions after an EMQX node restart, marking a significant enhancement for IoT applications requiring high reliability.
+
+::: warning Important Notice
+This feature is in the public alpha phase. We advise against its use in production environments containing critical data until further notice.
```
Comment (zmstone) [inline en_US/durability/durability_introduction.md]: beta?
```suggestion
This feature is in the public beta phase. We advise against its use in production environments containing critical data until further notice.
```

File: en_US/durability/durability_introduction.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+In EMQX, client sessions facilitate the management of client connections and states within the MQTT broker. Starting from version 5.5, EMQX offers two different implementations for managing the client session: regular client sessions and persistent client sessions.
+
+::: tip
+
+For more information on sessions, see [MQTT Persistent Session and Clean Session Explained](https://www.emqx.com/en/blog/mqtt-session).
+
+:::
+
+### Regular Client Sessions
```
Comment (zmstone) [inline en_US/durability/durability_introduction.md]: wonder if there is a better word than "regular".
maybe "ephemeral" ?

**Review by zmstone — COMMENTED** [review]
Thank you for the writeup.
We should probably document the current limitations too ?
e.g. shared subscription is not supported so far.
e.g. session states are only committed once every few seconds so the inflight state may not be reliable

**Comment by Meggielqk** [comment]
Also need to add more usage information from users' perspectives, such as configuration examples, how to verify the functions, etc. @wivwiv will add some inputs later.

File: en_US/durability/durability_introduction.md (hunk 4 above)
Comment (ieQu1) [reply en_US/durability/durability_introduction.md]: Removed this warning.

File: en_US/durability/introduction.md (hunk 1 above)
Comment (ieQu1) [reply en_US/durability/introduction.md]: @wivwiv Fixed in the latest commit.

File: en_US/durability/durability_introduction.md (hunk 5 above)
Comment (ieQu1) [reply en_US/durability/durability_introduction.md]: Fixed.

File: en_US/durability/durability_introduction.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+Storage encapsulates all data of a certain type, such as MQTT messages or MQTT sessions.
+
+#### Shard
+
+Messages are segregated by client and stored in shards based on the publisher's client ID. The number of shards is determined by [n_shards](./managing-replication.md#number-of-shards) configuration parameter during the initial startup of EMQX. A shard is also a unit of replication. Each shard is consistently replicated the number of times specified by `durable_storage.messages.replication_factor` across different nodes, ensuring identical message sets in each replica.
+
+#### Generation
+
+Messages within a shard are segmented into generations corresponding to specific time frames. New messages are written to the current generation, while previous generations are read-only. Old generations are deleted based on the `durable_sessions.message_retention_period` parameter.
```
Comment (ieQu1) [inline en_US/durability/durability_introduction.md]: @Meggielqk I think an important point got lost here, that _MQTT messages_ are cleaned up be deleting the generations containing them. Users don't care about generations, they care about the messages, so this part should be emphasized. From the user perspective generations only matter because they add granularity to the message retention.

File: en_US/durability/durability_introduction.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+- **Ephemeral**: ephemeral sessions exist only for the duration of the client's connection to EMQX. When a client with an ephemeral session disconnects, all session information, including subscriptions and undelivered messages, is discarded.
+
+The client session is considered persistent in two cases:
+
+1. For the clients using MQTT 5 protocol,
+   [Session Expiry Interval](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901048) property of `CONNECT` or `DISCONNECT` packet is set to a value greater than zero.
+
+2. For the clients using MQTT 3.* protocol,
+   [Clean Session](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718030) flag is set to 0,
+   and `mqtt.session_expiry_interval` configuration parameter is set to a value greater than 0.
```
Comment (Le9Wang) [inline en_US/durability/durability_introduction.md]: Seems mqtt.session_expiry_interval is a parameter in mqtt v5.

Comment (ieQu1) [reply en_US/durability/durability_introduction.md]: For V5 client this config doesn't override `Session-Expiry-Interval` property supplied by the client: https://github.com/emqx/emqx/blob/02097b577cdc4efca609deb06c0498eda05482d1/apps/emqx/src/emqx_channel.erl#L1537

File: en_US/durability/management.md [hunk 8]
```diff
@@ -1,91 +1,79 @@
...
 
-Storage layout determines how EMQX organizes data on disk.
-Layout used by the new [generations](./durability_introduction.html#generation) can be changed by setting `durable_storage.<DS>.layout.type` parameter.
-It does not affect the generations that has been already created.
-Currently, the following layouts are available:
+##### Configuration of `wildcard_optimized` Layout Type
 
-- `wildcard_optimized`
+The `wildcard_optimized` layout is aimed to optimize wildcard subscriptions across a vast array of topics. It achieves this by autonomously accumulating knowledge about topic structures over time. Leveraging a lightweight machine learning algorithm, it predicts the wildcard topic filters that clients are likely to subscribe to. Subsequently, it organizes these topics into a unified stream, facilitating efficient consumption in a single batch.
```
Comment (ieQu1) [inline en_US/durability/management.md]: > across a vast array of topics.

Not sure if I agree with the use of "vast array of topics" idiom in this context. Something like "matching large number of MQTT topics" is more precise.

Comment (ieQu1) [inline en_US/durability/management.md]: > facilitating efficient consumption in a single batch.

"facilitating efficient consumption in a single batch" -> "allowing efficient consumption in a single sweep". The original phrasing was imprecise too, because size of the batch is controlled by the client, layout type has nothing to do with that.

File: en_US/durability/managing-replication.md [hunk 9]
```diff
@@ -1,44 +1,44 @@
-# Replication Management
+# Manage Replication
 
-This document describes how to setup data replication and high availability for the EMQX durable storage.
-It is relevant when setting up a new EMQX cluster with session persistence enabled, or upgrading an existing cluster to enable session persistence.
+This guide provides instructions on configuring data replication and ensuring high availability for EMQX durable storage. This guide consists of instructions for two scenarios: setting up a new EMQX cluster with durable storage and upgrading an existing cluster to enable durable storage.
```
Comment (ieQu1) [inline en_US/durability/managing-replication.md:3]: > This guide provides

This document is not meant as a guide. It's meant to be a rather in-depth technical overview of the replication process, that answers most of the questions for the technically inclined users.

File: en_US/durability/management.md (hunk 8 above)
Comment (Meggielqk) [reply en_US/durability/management.md]: Thank you for the correction! I have updated it.

File: en_US/durability/durability_introduction.md (hunk 6 above)
Comment (Meggielqk) [reply en_US/durability/durability_introduction.md]: updated

File: en_US/durability/management.md [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
 
 ## Configuration Parameters
 
-Configuration for durable storage is divided into 2 main categories:
+Configuration for durable sessions is divided into 2 main categories:
 
-- `durable_sessions`: Contains settings related to MQTT clients' sessions, including how they consume data from durable storage and data retention parameters.
-- `durable_storage` Manages the settings of the durable storage system holding the MQTT message data.
+- `durable_sessions`: Contains settings related to MQTT clients' sessions, including how they consume data from durable sessions and data retention parameters.
+- `durable_storage` Manages the settings of the durable sessions system holding the MQTT message data.
```
Comment (ieQu1) [inline en_US/durability/management.md]: This is not correct. Please revert. Durable storage is a database-like application that does two things: it can save MQTT messages, and it can return saved MQTT messages. It has no concept of an MQTT session whatsoever.
This is why its configuration is separate from `durable_sessions`.

Comment (Meggielqk) [reply en_US/durability/management.md]: Sorry, it was mistakenly "search & replaced" due to the feature renaming. I have reverted the wording.

File: en_US/durability/management.md [hunk 11]
```diff
@@ -29,9 +29,9 @@ The following parameters can be overridden per [zone](../configuration/configura
 - `durable_sessions.idle_poll_interval`
 - `durable_sessions.renew_streams_interval`
 
-### Durable Storage Configuration
+### Durable Sessions Configuration
```
Comment (ieQu1) [inline en_US/durability/management.md]: Please revert.

File: en_US/durability/management.md [hunk 12]
```diff
@@ -29,9 +29,9 @@ The following parameters can be overridden per [zone](../configuration/configura
...
 
-The `<DS>` placeholder stands for "durable storage".  Currently, the available parameter for `<DS>` is `message`.
+The `<DS>` placeholder stands for "durable sessions".  Currently, the available parameter for `<DS>` is `message`.
```
Comment (ieQu1) [inline en_US/durability/management.md]: Please revert.

File: en_US/durability/management.md [hunk 13]
```diff
@@ -86,11 +86,11 @@ By default, the `epoch_bits` parameter is configured to 20 (~1 s), striking a ba
 
 ## CLI Commands
 
-The following CLI commands are available for managing the durable storage:
+The following CLI commands are available for managing the durable sessions:
```
Comment (ieQu1) [inline en_US/durability/management.md]: Please revert. These commands don't manage the sessions; they manage the storages.

File: en_US/durability/management.md [hunk 14]
```diff
@@ -125,15 +125,15 @@ This command output includes:
 
 - `THIS SITE`: ID of the site claimed by the local EMQX node.
 - `SITES`: List of all known sites, including EMQX node names and their statuses.
-- `SHARDS`: List of durable storage shards and site IDs where their replicas are located.
+- `SHARDS`: List of durable sessions shards and site IDs where their replicas are located.
```
Comment (ieQu1) [inline en_US/durability/management.md]: Please revert all changes from `storage` to `session` in this file. Management of durable sessions is done by standard EMQX routines (i.e. dashboard). This document talks specifically about _storages_. These terms are not interchangable.

File: en_US/durability/management.md [hunk 15]
```diff
@@ -191,7 +191,7 @@ The new section `REPLICA TRANSITIONS` lists pending operations. Once all operati
 
 ### `emqx_ctl ds join <DS> <Site>` / `emqx_ctl ds leave <DS> <Site>`
 
-These commands add or remove a site from the list of replicas of the durable storage. They are similar to the `set_replicas` command but update one site at a time.
+These commands add or remove a site from the list of replicas of the durable sessions. They are similar to the `set_replicas` command but update one site at a time.
```
Comment (ieQu1) [inline en_US/durability/management.md]: Please revert.

File: dir.yaml [hunk 16]
```diff
@@ -470,6 +469,12 @@
     - title_en: Command Line Interface
       title_cn: 命令行接口
       path: admin/cli
+    - title_en: Durable MQTT Sessions
```
Comment (wivwiv) [inline dir.yaml]: in the durability/durability_introduction file, the title is MQTT Durable Sessions
```suggestion
    - title_en: MQTT Durable Sessions
```

**Review by wivwiv — CHANGES_REQUESTED** [review]

**Review by wivwiv — DISMISSED** [review]

**Comment by wivwiv** [comment]
@ieQu1 Please change the target branch from release-5.6 to release-5.7.

**Review by wivwiv — DISMISSED** [review]

**Review by wivwiv — APPROVED** [review]
